Postgresql/legos/postgres_calculate_bloat/postgres_calculate_bloat.py

```diff
@@ -0,0 +1,163 @@
+    "JOIN pg_class ON tablename = relname "\
+    "JOIN pg_namespace ON relnamespace = pg_namespace.oid "\
+    "AND schemaname = nspname "\
+    "WHERE pg_class.relkind = 'r' "\
+    "), "\
+    "estimates_with_toast AS ( "\
+    "SELECT schemaname, tablename, "\
+    "TRUE as can_estimate, "\
+    "est_rows, "\
+    "table_bytes + ( coalesce(toast.relpages, 0) * bs ) as table_bytes, "\
+    "expected_bytes + ( ceil( coalesce(toast.reltuples, 0) / 4 ) * bs ) as expected_bytes "\
+    "FROM table_estimates LEFT OUTER JOIN pg_class as toast "\
+    "ON table_estimates.reltoastrelid = toast.oid "\
+    "AND toast.relkind = 't' "\
+    "), "\
+    "table_estimates_plus AS ( "\
+    "SELECT current_database() as databasename, "\
+    "schemaname, tablename, can_estimate, "\
+    "est_rows, "\
+    "CASE WHEN table_bytes > 0 "\
+    "THEN table_bytes::NUMERIC "\
+    "ELSE NULL::NUMERIC END "\
+    "AS table_bytes, "\
+    "CASE WHEN expected_bytes > 0 "\
+    "THEN expected_bytes::NUMERIC "\
+    "ELSE NULL::NUMERIC END "\
+    "AS expected_bytes, "\
+    "CASE WHEN expected_bytes > 0 AND table_bytes > 0 "\
+    "AND expected_bytes <= table_bytes "\
+    "THEN (table_bytes - expected_bytes)::NUMERIC "\
+    "ELSE 0::NUMERIC END AS bloat_bytes "\
+    "FROM estimates_with_toast "\
+    "UNION ALL "\
+    "SELECT current_database() as databasename, "\
+    "table_schema, table_name, FALSE, "\
+    "est_rows, table_size, "\
+    "NULL::NUMERIC, NULL::NUMERIC "\
+    "FROM no_stats "\
+    "), "\
+    "bloat_data AS ( "\
+    "select current_database() as databasename, "\
+    "schemaname, tablename, can_estimate, "\
+    "table_bytes, round(table_bytes/(1024^2)::NUMERIC,3) as table_mb, "\
+    "expected_bytes, round(expected_bytes/(1024^2)::NUMERIC,3) as expected_mb, "\
+    "round(bloat_bytes*100/table_bytes) as pct_bloat, "\
+    "round(bloat_bytes/(1024::NUMERIC^2),2) as mb_bloat, "\
+    "table_bytes, expected_bytes, est_rows "\
+    "FROM table_estimates_plus "\
+    ") "\
+    "SELECT databasename, schemaname, tablename, "\
+    "can_estimate, "\
+    "est_rows, "\
+    "pct_bloat, mb_bloat, "\
+    "table_mb "\
+    "FROM bloat_data "\
+    "ORDER BY pct_bloat DESC; "
+
+    cur = handle.cursor()
+    cur.execute(query)
+    output = []
+    res = cur.fetchall()
+    data = []
+    for records in res:
+        result = {
+            "database_name": records[0],
+            "schema_name": records[1],
+            "table_name": records[2],
+            "can_estimate": records[3],
+            "live_rows_count": records[4],
+            "pct_bloat": records[5],
+            "mb_bloat": records[6],
+            "table_mb": records[7]
+        }
+        output.append(result)
+        data.append([records[0], records[1], records[2], records[5], records[6]])
+
+    if len(res) > 0:
+        headers = ["db_name", "schema_name", "table_name", "pct_bloat", "mb_bloat"]
+        output = tabulate(data, headers=headers, tablefmt="grid")
```

Comment:
@shloka-bhalgat-unskript this doesnt look right, the printer function of the lego should do the tabulate. If  i need to use the output of the above lego in the next lego, how will i use it? This lego should return the LIST, which is data.

Reply:
Fixed @amit-chandak-unskript